src/btree/bt_cursor.c

```diff
@@ -842,49 +840,69 @@ __wt_btcur_next_random(WT_CURSOR_BTREE *cbt)
+		 * unbalanced trees (sometimes, very, very unbalanced trees),
+		 * and in all WiredTiger releases, internal and leaf page sizes
+		 * aren't uniform. So, we skip past 1% of the leaf pages in the
+		 * tree between each random key return to compensate.
+		 *
+		 * Deciding how many leaf pages is roughly 1% of the file is not
+		 * simple. Use the underlying file size divided by its block
+		 * allocation size as our guess of leaf pages in the file (this
+		 * can be entirely wrong, as it depends on how many pages are in
+		 * this particular checkpoint, how large the leaf and internal
+		 * pages really are, and other factors). Then, divide that value
+		 * by 100 to get 1%, and increment the final result to make sure
+		 * tiny files don't leave us with a skip value of 0.
+		 *
+		 * !!!
+		 * Ideally, the number would be prime to avoid restart issues.
 		 */
-		if (cbt->rand_leaf_skip == 0) {
-			cbt->rand_leaf_skip = (uint32_t)
-			    (btree->bm->block->fh->size / btree->allocsize);
-			/* Walk 1% of pages each time, adjust for tiny files. */
-			cbt->rand_leaf_skip /= 100;
-			++cbt->rand_leaf_skip;
-			printf("Skipping %u leaf pages each random op\n",
-			    cbt->rand_leaf_skip);
-		}
-
-		/* Walk through the current parent. */
-		leaf = 0;
-		parent_ref = cbt->ref->home->pg_intl_parent_ref;
-		WT_RET(__wt_page_swap(session, cbt->ref, parent_ref, 0));
-		cbt->ref = parent_ref;
-
-		while (leaf < cbt->rand_leaf_skip) {
-			parent = cbt->ref->page;
-			WT_INTL_FOREACH_BEGIN(session, parent, child) {
-				if (__wt_ref_is_leaf(session, child) &&
-				    ++leaf == cbt->rand_leaf_skip) {
-					WT_RET(__wt_page_swap(
-					    session, cbt->ref, child, 0));
-					cbt->ref = child;
-					goto leaf;
-				}
-			} WT_INTL_FOREACH_END;
-			do {
-				WT_RET(__wt_tree_walk(session,
-				    &cbt->ref, NULL, WT_READ_SKIP_LEAF));
-			} while (cbt->ref == NULL);
-		}
+		cbt->rand_leaf_skip = (uint32_t)
+		    ((btree->bm->block->fh->size / btree->allocsize) / 100) + 1;
+#if 0
+		printf("Skipping %u leaf pages each random op\n",
+		    cbt->rand_leaf_skip);
+#endif
+		/*
+		 * Choose a leaf page from the tree.
+		 */
+		WT_ERR(__cursor_func_init(cbt, true));
+		WT_WITH_PAGE_INDEX(
+		    session, ret = __wt_row_random_descent(session, cbt));
+		WT_ERR(ret);
+	} else {
+		/*
+		 * Read through the tree, skipping leaf pages. Be cautious about
+		 * the skip count: if the last leaf page skipped was also the
+		 * last leaf page in the tree, it may be set to zero on return
+		 * with the end-of-walk condition.
+		 *
+		 * Pages read for data sampling aren't "useful"; don't update
+		 * the read generation of pages already in memory, and if a page
+		 * is read, set its generation to a low value so it is evicted
+		 * quickly.
+		 */
+		for (skip = cbt->rand_leaf_skip; cbt->ref == NULL || skip > 0;)
+			WT_ERR(__wt_tree_walk(session, &cbt->ref, &skip,
+			    WT_READ_NO_GEN | WT_READ_SKIP_INTL |
+			    WT_READ_SKIP_LEAF | WT_READ_WONT_NEED));
```

Comment:
@michaelcahill: my inclination is to not rename as `WT_READ_SAMPLE`, if at a future date we want to skip leaves for other reasons, that will end up being hard to work with.

We could add a parameter that set the number of leaves to skip (and it would be nice it if had an out-of-band value that caused all leaves to be skipped).